### PR TITLE
conform() with Sets added too many items

### PR DESCRIPTION
**Pull request by mattbishop**
js.spec's Set handling is broken WRT conform() output. The returned Set includes an extra 'length' field.
This PR fixes how conform assembles the return value. It upgrades the unit testing to chai 4 to gain improved Set assertions.

### CODE REVIEW

File: test/spec/collection.test.js [hunk 1]
```diff
@@ -6,6 +6,12 @@ import { explainData } from "../../index";
 import { invalid, count, minCount, maxCount } from "../../lib/symbols";
 
 describe("collection", () => {
+  describe("factory", () => {
```
Comment (mattbishop) [inline test/spec/collection.test.js]: Coveralls was complaining about reduced test coverage so I added a new test to increase coverage.

File: lib/spec/collection.js [hunk 2]
```diff
@@ -34,13 +34,13 @@ export class Collection extends Spec {
...
     }
 
-    const ret = [];
-    for (const [key, val] of value.entries()) {
-      const conformed = dt(this.options.spec, val);
-      if (conformed === invalid) {
-        return invalid;
-      }
-      ret[key] = conformed;
+    const spreadValue = p.array(value) ? value : [...value];
```
Comment (mattbishop) [inline lib/spec/collection.js]: for / of was adding in a `length` key, which was the source of this bug. Using spread operator converts the Set to an array with simpler map / filter functions.

File: package.json [hunk 3]
```diff
@@ -36,11 +36,11 @@
     "babel-plugin-transform-object-rest-spread": "^6.23.0",
     "babel-preset-es2015": "^6.18.0",
     "babel-register": "^6.18.0",
-    "chai": "^3.5.0",
-    "coveralls": "^2.11.14",
-    "eslint": "^4.4.1",
-    "eslint-plugin-prettier": "^2.1.2",
-    "mocha": "^3.1.2",
+    "chai": "^4.1.1",
```
Comment (mattbishop) [inline package.json]: Set testing was broken in old unit testing libs. I upgraded these together to the latest version to be safe.

File: test/spec/collection.test.js [hunk 4]
```diff
@@ -206,13 +212,15 @@ describe("collection", () => {
...
-        new Set([1, 2])
-      );
+      expect(
+        spec.conform([1, 2]),
+        "array - count, happy"
+      ).to.contain.ordered.members([1, 2]);
+
+      expect(spec.conform(new Set([1, 2])), "set - count, happy").to.be
+        .a("Set")
+        .that.contains(1, 2)
```
Comment (mattbishop) [inline test/spec/collection.test.js]: `deep.equal` failed for Sets. It considered the two Sets equal, even though one has 3 elements and the other has 2.

File: package.json (hunk 3 above)
Comment (arichiardi) [inline package.json]: :+1: Related -> https://github.com/chaijs/deep-eql/issues/4

File: test/spec/collection.test.js (hunk 4 above)
Comment (arichiardi) [inline test/spec/collection.test.js]: :+1: Related -> https://github.com/chaijs/deep-eql/issues/4

**Comment by arichiardi** [comment]
LGTM

**Comment by mattbishop** [comment]
@prayerslayer any comments on this PR?

**Comment by arichiardi** [comment]
Will push a new version as well

**Pull request merged**